### PR TITLE
Adopting `ruff` formatter

### LINKED ISSUE #7914: Use ruff for formatting
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

Can I implement this upgrade in the CI? What package is currently being used?

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [ ] [Data validation/parsing](https://docs.pydantic.dev/concepts/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/concepts/serialization/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/concepts/json_schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/concepts/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/concepts/config/)
- [ ] [Field Types](https://docs.pydantic.dev/api/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/concepts/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/concepts/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/concepts/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

### PR DESCRIPTION
**Pull request by Luca-Blight**
### Change Summary

Ruff is now the sole linter 👍 
Files Altered: ['Makefile', 'docs/contributing.md', 'pyproject.toml', 'pdm.lock' ]

Notes: 

For pyproject.toml file,  I removed the last element('E501') of line 180 given black's removal but I'm not totally sure what this line is for. 

### Related Issue Number

Fix #7914

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

**Comment by Luca-Blight** [comment]
I suppose I could run Ruff local to fix the linting issues...what do you think @sydney-runkle?

**Comment by sydney-runkle** [comment]
Hi @Luca-Blight,

Looks good! Yeah, go ahead and run the linter locally to fix the failing checks. I'll take another look after you do that.

Thanks 🌟

File: pyproject.toml
Comment (sydney-runkle) [inline pyproject.toml:93]: I think we might want to pin a version here, `ruff==0.1.2`

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
Thanks for running the linter locally. I suggested a couple of very minor changes 👍

File: Makefile
Comment (sydney-runkle) [inline Makefile:28]:
```suggestion
	pdm run ruff format $(sources)
```

**Review by sydney-runkle — CHANGES_REQUESTED** [review]

File: pydantic/_internal/_decorators.py
Comment (sydney-runkle) [inline pydantic/_internal/_decorators.py:156]: This seems odd to me, looking at ruff docs now to learn more...

File: pydantic/_internal/_model_construction.py
Comment (sydney-runkle) [inline pydantic/_internal/_model_construction.py:373]: This change also seems odd to me, as it's not changing the last line. I wonder what happens when we use:
```
[tool.ruff.flake8-quotes]
multiline-quotes = 'double'
```
In the `pyproject.toml`

Comment (sydney-runkle) [reply pydantic/_internal/_model_construction.py:373]: Actually, it should be `'double'` by default, so I don't think that'll change anything. This modification still strikes me as odd though.

Comment (Luca-Blight) [reply pydantic/_internal/_model_construction.py:373]: Yea good eye there. Let me try the change out and see what happens.

Comment (Luca-Blight) [reply pydantic/_internal/_model_construction.py:373]: Btw, your diligence is impressive and I appreciate it.

Comment (Luca-Blight) [reply pydantic/_internal/_model_construction.py:373]: It looks like
```
[tool.ruff.flake8-quotes]
multiline-quotes = 'double'
```

is already implemented on line 179:

```
[tool.ruff]
line-length = 120
extend-select = ['Q', 'RUF100', 'C90', 'UP', 'I', 'D', 'T']
extend-ignore = ['D105', 'D107', 'D205', 'D415']
flake8-quotes = {inline-quotes = 'single', multiline-quotes = 'double'}
mccabe = { max-complexity = 14 }
isort = { known-first-party = ['pydantic', 'tests'] }
target-version = "py37"
extend-exclude = ['pydantic/v1', 'tests/mypy/outputs']
```

Comment (Luca-Blight) [reply pydantic/_internal/_model_construction.py:373]: If I attempt to add the first code block it naturally raise a conflict error.

Comment (sydney-runkle) [reply pydantic/_internal/_model_construction.py:373]: Good catch. I'll leave this as unresolved for now given that I'm still curious about the change the `ruff` formatter made, but I don't think it's a problem in terms of the `flake8-quotes` configuration.

File: Makefile
Comment (Luca-Blight) [reply Makefile:28]: Change for line 32 as well?

File: tests/test_networks.py
Comment (sydney-runkle) [inline tests/test_networks.py:808]: Another odd change

File: tests/test_private_attributes.py
Comment (sydney-runkle) [inline tests/test_private_attributes.py:292]: Clearly it's having a hard time with this as well 😆

File: Makefile
Comment (sydney-runkle) [reply Makefile:28]: @Luca-Blight I think we want to leave line 32 as is, as you call the [ruff linter](https://docs.astral.sh/ruff/linter/) with slightly different syntax compared to the [ruff formatter](https://docs.astral.sh/ruff/formatter/)

File: tests/test_private_attributes.py
Comment (Luca-Blight) [reply tests/test_private_attributes.py:292]: Yea that is weird. I wonder what it's tripping on...

File: pydantic/_internal/_decorators.py
Comment (sydney-runkle) [reply pydantic/_internal/_decorators.py:156]: Aha, this is relevant: https://github.com/astral-sh/ruff/issues/8226

Comment (sydney-runkle) [reply pydantic/_internal/_decorators.py:156]: @Luca-Blight, how many of these issues are fixed if you pull down the latest version of `ruff`?

Comment (Luca-Blight) [reply pydantic/_internal/_decorators.py:156]: I'm not sure it will cause issue but it certainly is funny looking. I wonder if they have a fix in for it, it will probably not be atleast until the next version.

Comment (Luca-Blight) [reply pydantic/_internal/_decorators.py:156]: Nice job mentioning the issue in their thread.

Comment (sydney-runkle) [reply pydantic/_internal/_decorators.py:156]: Yeah looks like they're working hard to get some fixes across the line. I can pull down the latest version tmrw morning and see what's resolved.

**Comment by sydney-runkle** [comment]
@Luca-Blight,

Thanks for your work on this 💯. Hopefully we can resolve the few odd formatting patterns that we're seeing tomorrow and get this across the line 😄.

File: pydantic/_internal/_decorators.py
Comment (Luca-Blight) [reply pydantic/_internal/_decorators.py:156]: Fwiw, I shared this thread with the founding/lead engineer. Perhaps they can expedite it the resolution if not available tomorrow.

File: tests/test_networks.py
Comment (MichaReiser) [reply tests/test_networks.py:808]: Ruff's formatter removes unnecessary escape sequences and picks the quote style that leads to the fewest escape (with the motivation that escapes reduce readability). Also see https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#strings

File: pydantic/_internal/_decorators.py
Comment (charliermarsh) [reply pydantic/_internal/_decorators.py:156]: I think we're likely to fix this shortly. (There's a PR up, just trying to make some decisions around the broader annotation formatting before merging.)

File: pydantic/_internal/_model_construction.py
Comment (charliermarsh) [reply pydantic/_internal/_model_construction.py:373]: This strikes me as a bug -- we should be using double quotes for multi-line strings regardless of the formatter quote setting. Will look into it shortly and report back.

Comment (MichaReiser) [reply pydantic/_internal/_model_construction.py:373]: The last line doesn't get changed because changing the quote style would require escaping the `'`. See my other comment below.

File: tests/test_networks.py
Comment (charliermarsh) [reply tests/test_networks.py:808]: I believe the backslash is unnecessary here -- that is, the string is identical with and without the backslash:

```python
>>> '\"@example.com'
'"@example.com'
```

File: tests/test_private_attributes.py
Comment (charliermarsh) [reply tests/test_private_attributes.py:292]: So, what's happening here is that it's trying to convert each segment to single quotes (as per `quote-style = 'single'`). It's succeeding for the first segment in this implicit concatenation, but it avoids changing the second segment, because that segment itself _contains_ a single quote (and so that single quote would need to be escaped).

Arguably we should be making that decision across the _entire_ implicit concatenation (i.e., looking across all segments) to avoid these mixed quotes. I'll file an issue and we'll discuss.

Comment (MichaReiser) [reply tests/test_private_attributes.py:292]: Ruff changes the quotes of the first string but not of the second because it would require escaping the `'`.  Is it unexpected that Ruff uses different quotes for different parts of the implicit concatenated string? Would you prefer if it would change the quotes consistently for the entire implicit concatenated string?

File: pydantic/_internal/_decorators.py
Comment (MichaReiser) [reply pydantic/_internal/_decorators.py:156]: For reference https://github.com/astral-sh/ruff/pull/8233

File: pydantic/_internal/_model_construction.py
Comment (charliermarsh) [reply pydantic/_internal/_model_construction.py:373]: Sorry, as usual, Micha is right -- this is an implicit concatenation, not a multiline string. So it's trying to use single quotes for each segment, and then skipping those that _contain_ a single quote.

File: pydantic/_internal/_decorators.py
Comment (sydney-runkle) [reply pydantic/_internal/_decorators.py:156]: @charliermarsh @MichaReiser thanks a ton for your work on this awesome tool. I'll keep an eye on that issue with the fix 😄

File: pydantic/_internal/_model_construction.py
Comment (sydney-runkle) [reply pydantic/_internal/_model_construction.py:373]: Ah, I see. This happened a few times with implicit concatenations, and I thought the pattern was odd at first glance bc it always seemed to be skipping the last line, but that seems to always be where we have a single quote based on the nature of many of our strings. Thanks!

File: tests/test_networks.py
Comment (sydney-runkle) [reply tests/test_networks.py:808]: That's fair. I suppose we should just remove the test case then if it's identical to the one directly above.

File: tests/test_private_attributes.py
Comment (sydney-runkle) [reply tests/test_private_attributes.py:292]: Ah, thanks for the explanation. Clearly I missed the pattern of escaping `'` that was at the root of many of my questions.

I do feel like changing quotes consistently for an entire implicit concatenated string might look a bit cleaner, but I'm not adamant about the change. Excited to hear what others think. 

@charliermarsh, thanks for filing an issue 👍

**Comment by sydney-runkle** [comment]
@charliermarsh @MichaReiser,

Thanks for your super prompt support. Looks like we'll have this across the line in no time 🚀 !

File: tests/test_networks.py
Comment (charliermarsh) [reply tests/test_networks.py:808]: (Sorry, I think Micha and I just hit a race condition here, didn't mean to overload you with my follow-up comment!)

File: pydantic/_internal/_decorators.py
Comment (sydney-runkle) [reply pydantic/_internal/_decorators.py:156]: @charliermarsh I see that issue has been resolved. Do you know when you all play to make another release including that fix?

File: tests/test_private_attributes.py
Comment (sydney-runkle) [reply tests/test_private_attributes.py:292]: @charliermarsh could you link the issue here? I may have just missed it, but didn't see it at a first glance

File: tests/test_networks.py
Comment (sydney-runkle) [reply tests/test_networks.py]: No worries at all. I removed the redundant test case and am marking this as resolved :)

File: pydantic/_internal/_decorators.py
Comment (charliermarsh) [reply pydantic/_internal/_decorators.py:156]: I’m planning to cut a release today :)

Comment (sydney-runkle) [reply pydantic/_internal/_decorators.py:156]: @Luca-Blight, we'll want to revert changes that resemble the example above

Comment (charliermarsh) [reply pydantic/_internal/_decorators.py:156]: Out now: https://github.com/astral-sh/ruff/releases/tag/v0.1.3. I also confirmed locally that this change is no longer applied to Pydantic (though you'll want to revert it here -- like Black, if you use this formatting with the parentheses, we do _retain_ it).

**Review by sydney-runkle — APPROVED** [review]
LGTM 👍

The only thing I'm still seeing is the implicit concatenation changes that we discussed already, which I don't think we have to address in this PR, but can continue discussion with `ruff` and change this down the line.

@Luca-Blight, great work!

File: pydantic/_internal/_decorators.py
Comment (sydney-runkle) [reply pydantic/_internal/_decorators.py]: Awesome, thanks! We reverted the changes 😄 and it's working as expected now 🌟

**Comment by Luca-Blight** [comment]
> LGTM 👍
> 
> The only thing I'm still seeing is the implicit concatenation changes that we discussed already, which I don't think we have to address in this PR, but can continue discussion with `ruff` and change this down the line.
> 
> @Luca-Blight, great work!

I noticed some letter swapping in some of the files such as rf -> fr and br -> rb. 
There are several examples in the test files and color.py. Do you see that as an issue?

**Comment by sydney-runkle** [comment]
> I noticed some letter swapping in some of the files such as rf -> fr and br -> rb. There are several examples in the test files and color.py. Do you see that as an issue?

Yeah I saw that as well. I don't think it's an issue. @adriangb or @dmontagu, do you think otherwise?

**Comment by adriangb** [comment]
I see it's changing `rf'a string'` to `fr'a string'`. I agree it doesn't matter.

File: tests/test_networks.py
Comment (dmontagu) [inline tests/test_networks.py:808]: was this change on purpose? It seems to not be a formatting change

File: tests/test_plugins.py
Comment (dmontagu) [inline tests/test_plugins.py:287]: @charliermarsh if you are looking for feedback, I'd think that it would be preferable to use single quotes on the outer level when single quotes are preferred and the string contains both single and double quotes. But I definitely don't care enough to hold off on merging just for this issue 😄

Comment (charliermarsh) [reply tests/test_plugins.py:287]: Yeah I agree — this one struck me as odd. Gonna look into it.

File: tests/test_networks.py
Comment (sydney-runkle) [reply tests/test_networks.py:808]: Yeah, I removed this test case bc `ruff` was removing the `\`. This case and the one directly above are identical.

Comment (charliermarsh) [reply tests/test_networks.py:808]: I believe the string formatter removes unnecessary backslashes. (The backslash here doesn’t have an effect.)

**Review by dmontagu — APPROVED** [review]

File: tests/test_plugins.py
Comment (charliermarsh) [reply tests/test_plugins.py:287]: https://github.com/astral-sh/ruff/issues/8264

File: tests/test_private_attributes.py
Comment (charliermarsh) [reply tests/test_private_attributes.py:292]: Apologies -- just filed it here: https://github.com/astral-sh/ruff/issues/8265

**Pull request merged**
